Senedd ID changes

chris48s opened the pull request:
Refs https://app.asana.com/1/1204880536137786/project/1204880927741389/task/1210252700116907?focus=true

Lets make the first baby steps towards this. This will be part one of many.

The first decision I have made is that the way we should implement this at the library layer is we say that subtypes are **optional** for Senedd elections, but we don't define when they are/are not valid here.

So in the same way that this library will let you create/validate

- `sp.c.shetland-islands.1980-08-29` (there was no Scottish Parliament in 1980) or
- `parl.this-constituency-does-not-exist.2017-03-02` (invalid division name)

the library will allow you to create/validate

- `senedd.r.mid-and-west-wales.2026-05-07` or
- `senedd.pen-y-bont-bro-morgannwg.2016-05-05`

(both equally invalid)

As with the other cases, it is up to Every Election's ID builder to plug the right values into it at the right times.

So then having made that decision, I have gone ahead and modified the test suite with test cases based on that assumption. Note that I have not yet changed any code or functionality, so the tests are now failing. This is expected.

Assuming we agree that the changes to the test suite formalise the library behaviour we want then the next step is: Change the code so that the tests pass.

chris48s committed:
add failing test cases for optional subtype layer in senedd ids

symroe commented:
I think this is broadly sensible as a way to get started. 

The reason we didn't add division name validation is that we didn't want to export a list of all divisions into this library. I think there's not really a good case for doing this.

However, I do see there might be a good case for this library to do some other sorts of validation in the future. This is a good example where the library knowing about the start/end dates of the higher level concepts seems to make sense. 

As we don't currently do this though, I don't think now is the time to start. The main uses for this, outside of EE making new IDs, is validating IDs. Your suggested changes will allow this to work, so I think it's fine.

GeoWill commented on tests/test_validator.py:
```diff
@@ -61,17 +67,25 @@ def test_invalid_ids(self):
         )
+        self.assertTrue(
+            validate(
+                "senedd.pen-y-bont-bro-morgannwg.something-else.2026-05-07"
```
This seems to be a separate change? In this case I think the only thing that goes between `division` and `date` is an optional `by`.
I guess the other way of reading the id is as a division called `something-else` with a subtype of `pen-y-bont-bro-morgannwg`. But then shouldn't this be invalid, because that wouldn't be a valid subtype.

chris48s committed:
fix copy+paste error

chris48s committed:
allow senedd elections to have optional subtypes

chris48s commented on uk_election_ids/datapackage.py:
```diff
@@ -61,6 +64,7 @@
                 "can_have_divs": False,
             },
         ],
+        "subtypes_required": True,
```
Changing `ELECTION_TYPES` shouldn't really be a breaking change as it is undocumented internals

However in practice, these have become load-bearing undocumented internals. Yay. Changing this has knock-on effects we may need to account for in some, all, or none of:

## Every Election
- https://github.com/DemocracyClub/EveryElection/blob/master/every_election/apps/elections/constants.py
- https://github.com/DemocracyClub/EveryElection/blob/master/every_election/apps/elections/utils.py
- https://github.com/DemocracyClub/EveryElection/blob/master/every_election/apps/elections/management/commands/add_election_types.py
- https://github.com/DemocracyClub/EveryElection/blob/master/every_election/apps/elections/migrations/0003_initial_data.py
- https://github.com/DemocracyClub/EveryElection/blob/master/every_election/apps/elections/tests/conftest.py
- https://github.com/DemocracyClub/EveryElection/blob/master/every_election/apps/elections/views/general.py

## Response Builder
- https://github.com/DemocracyClub/dc_response_builder/blob/b3708ffa129ba1a40cded96455eff963432ca5a4/tests/test_builders.py#L4

So I've tried to do this in a way that changes as little as possible.

**TODO**: review all those

chris48s commented on uk_election_ids/election_ids.py:
```diff
@@ -429,7 +434,11 @@ def from_id(cls, identifier: str) -> "IdBuilder":
                 id_parts.pop(-1)
                 builder = builder.with_contest_type("by")
-            if builder.spec.subtypes:
+            if (builder.spec.subtypes and builder.spec.subtypes_required) or (
+                builder.spec.subtypes
+                and not builder.spec.subtypes_required
+                and len(id_parts[0]) == 1
+            ):
                 subtype = id_parts.pop(0)
                 builder = builder.with_subtype(subtype)
```
There's an interesting thing here.
If we want to reject `senedd.x.2021-05-06` as "invalid subtype"
we have to make an assumption
i.e: why is 'x' an invalid subtype and not the name of a division

The assumption I've made here is that if subtypes are optional and the next clause of the string is one character then we assume it is trying to be a subtype
This is fine as long as:
- We never want to have subtypes longer than one character
- There is never an electoral division with a name that is only one character long under an election type with optional subtypes

I think we are probably relatively safe there

However if we didn't want to box ourselves in here, we could make this:

```py
if (builder.spec.subtypes and builder.spec.subtypes_required) or (
    builder.spec.subtypes
    and not builder.spec.subtypes_required
    and id_parts[0] in builder.spec.subtypes
):
    subtype = id_parts.pop(0)
    builder = builder.with_subtype(subtype)
```

here. The tradeoff being that `senedd.x.2021-05-06` is considered a valid with the division slug being 'x'.

I'm kinda fine with either of those tbh

chris48s replied:
Note in the second scenario, we can still reject an invalid subtype for election types with _required_ subtypes (SP, GLA)

chris48s replied:
I would like feedback on this as part of the review

symroe replied:
Here's maybe another way to look at this problem:

We don't want, in this library to have a full enumeration of all possible values for all orgs, divisions, etc. We make an exception to election types and sub-types basically because they're tiny enumerations (and it's a useful thing to do).

I think this library should continue with this approach, but that's not to say there is never value in validating that the values _are_ correct. This validation could include checking a division name for a given date, etc.

To save trying to solve too many problems here, if we want to add more validation I feel like we should all consuming applications to extend the validation they perform. This would allow EE to perform lookups in its database at validation time to confirm that an ID _is_ actually valid.

I don't think we need to do this now, but I think conceptually it's a better approach than trying to make too many assumptions here. 

To that end, I think I have a weak preference for `id_parts[0] in builder.spec.subtypes` (the second option), and accepting that without an enumeration of divisions we will just accept `x`. 

As I say, this isn't a strong opinion and I'm up for talking about it if I've missed something.

chris48s committed:
remove default voting systems from pcc and mayor election types

chris48s committed:
update senedd voting systems definitions

chris48s committed:
define AMS on the relevant subtype

chris48s committed:
delete test and define this constraint at the schema level

chris48s committed:
add more tests for voting systems data

chris48s committed:
throw ValueError if no match in matcher classes

chris48s committed:
add tests for cases where it doesn't make sense to return voting system

It doesn't make sense to return a voting system for
top-level IDs (date groups) in the following cases:

- local.date (this could contain elections in Scot/NI and Eng/Wales)
- naw.date, gla.date, sp.date (AMS)
- senedd.date before 2026-05-06 (also AMS)

chris48s committed:
update AMS name/description

chris48s commented on uk_election_ids/data/voting_systems.json:
```diff
@@ -61,29 +61,43 @@
     },
     "senedd.c": {
-      "default": "FPTP"
+      "dates": {
+        ":2026-05-06": {
+          "default": "FPTP"
+        }
+      }
+    },
+    "senedd.r": {
+      "dates": {
+        ":2026-05-06": {
+          "default": "AMS"
+        }
+      }
     },
     "senedd": {
-      "default": "AMS"
+      "dates": {
+        "2026-05-07:": {
+          "default": "PR-CL"
+        }
+      }
     },
     "sp.c": {
       "default": "FPTP"
     },
-    "sp": {
+    "sp.r": {
       "default": "AMS"
```
Upshot of that conversation:
We're not going to set any voting system on the top-level election for AMS elections (so sp.date, gla,date, etc). We're going to change the name/description of the slug "AMS" to refer to just the proportional/party list bit of the AMS election. So an AMS election is represented as:

```
"sp.c": {
  "default": "FPTP"
},
"sp.r": {
  "default": "AMS"
}
```

in the data. Note there is a bit of nuance coming up on what we do with the top-level election...

chris48s commented on tests/test_voting_systems.py:
```diff
@@ -1,37 +1,124 @@
+            "FPTP",
+        )
 
+    def test_senedd_with_subtypes(self):
+        for id_ in [
+            "senedd.r.mid-and-west-wales.2021-05-06",
+            "senedd.r.2021-05-06",
+        ]:
+            self.assertEqual(
+                VotingSystemMatcher(id_).get_voting_system(),
+                "AMS",
+            )
+        for id_ in [
+            "senedd.c.swansea-east.2021-05-06",
+            "senedd.c.2021-05-06",
+        ]:
+            self.assertEqual(
+                VotingSystemMatcher(id_).get_voting_system(),
+                "FPTP",
+            )
```
A pattern we have a lot in this codebase is:
We want to test lots of slightly different valid input generate an expected output.
What we've tended to do is write one test that makes a bunch of different assertions rather than writing an individual named test for each one. A pattern I really like for this is parametrised tests. So you can write stuff like this
https://github.com/chris48s/stage-left/blob/c3b096b44831f863e11829cf5f2e4974c18a4f30/tests/test_date_parsers.py#L108-L135
Given the type of tests we write, I think that would be a really good fit for this project but given we're not using pytest I can't just start using it. I think "migrate the entire test suite" to pytest is not a good increase in scope for this PR :smile: so for the moment I'm just going to carry on writing tests that do more than one thing. I do think there is some mileage in it though.

symroe replied:
I'd like to be using `pytest` everywhere ideally, but agree this isn't the time and it's hard to justify the time to do it. FWIW, we can use `subTest` here to get to almost the same point without re-building everything. Doing that now (soon) might help with a later refactor too.

chris48s replied:
Subtest seems like a good improvement to make :+1: 
I have just pushed fafa822 which moves to using subtest in `test_validator.py` and `test_voting_systems.py` which are the two main test files I am modifying in this PR. I am not going to refactor the entire test suite here, but I'll keep this principle in mind as I work on further PRs.

chris48s commented:
OK. Last post.. follow up work

- https://github.com/DemocracyClub/uk-election-ids/pull/21#discussion_r2169185909
- Having made a decision on https://github.com/DemocracyClub/uk-election-ids/pull/21#discussion_r2177914394 update any code (mainly EE, but maybe other code?) that consumes `VOTING_SYSTEMS` or `VotingSystemMatcher`
- Update EE: Don't set a voting system when it doesn't make sense
- Fix historic data where we have done that
- Review any impact on WCIVF:
    - https://github.com/DemocracyClub/WhoCanIVoteFor/blob/5f96a03575f75816de64db83589b4e43ac9a18ae/wcivf/apps/elections/templates/elections/ams.html#L5
    - https://github.com/DemocracyClub/WhoCanIVoteFor/blob/5f96a03575f75816de64db83589b4e43ac9a18ae/wcivf/apps/elections/templates/elections/includes/_how-to-vote.html#L7

..and then we're back to the original point, which was Senedd stuff :smile:

symroe commented on tests/data_schemas.py:
```diff
@@ -13,8 +13,19 @@ class VotingSystem(BaseModel):
     default: Optional[str]
-    nations: Optional[dict]
-    dates: Optional[dict]
+    nations: Optional[Dict[str, "SingleVotingSystemRule"]]
+    dates: Optional[Dict[str, "SingleVotingSystemRule"]]
+
+    @root_validator
+    def at_least_one_required(cls, values):
+        if not any(
+            values.get(key) is not None
+            for key in ("default", "nations", "dates")
+        ):
+            raise ValueError(
+                "At least one of 'default', 'nations', or 'dates' must be present"
+            )
+        return values
```
Small code point: we might need to [call `update_forward_refs`](https://docs.pydantic.dev/1.10/usage/postponed_annotations/) in the tests.

chris48s replied:
Thanks for the link.

I tried making the input file invalid locally, and the schema seems to reject invalid inputs as expected without a call to `SingleVotingSystemRule.update_forward_refs()`. Adding a call to `SingleVotingSystemRule.update_forward_refs()` seems to neither break nor fix anything. Having read the article, I think this is because I used a string type hint instead of `typing.ForwardRef` https://docs.pydantic.dev/1.10/usage/postponed_annotations/#:~:text=In%20cases%20where%20the%20referenced%20type%20is%20not%20yet%20defined%2C%20ForwardRef%20can%20be%20used%20(although%20referencing%20the%20type%20directly%20or%20by%20its%20string%20is%20a%20simpler%20solution%20in%20the%20case%20of%20self%2Dreferencing%20models).

Do you agree the schema changes cover the same things we were looking for in the test I removed?

chris48s replied:
Looking into this did make me realise I wanted to edit the schema slightly so that we get a useful error if validating a sub-object against `SingleVotingSystemRule` fails though. See 454b769

chris48s committed:
only pop subtype if it is a known subtype

chris48s committed:
only try to recurse if 'defaults' exists

this means we see a meaningful error if
SingleVotingSystemRule throws ValueError

chris48s committed:
move to using subtests in test_validator and test_voting_systems
